keps/sig-storage/20200127-recover-from-resize.md

```diff
@@ -0,0 +1,153 @@
+- [Graduation Criteria](#graduation-criteria)
+  - [Test Plan](#test-plan)
+  - [Monitoring](#monitoring)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+A user may expand a PersistentVolumeClaim(PVC) to a size which may not be supported by
+underlying storage provider. In which case - typically expansion controller forever tries to
+expand the volume and keeps failing. We want to make it easier for users to recover from
+expansion failures, so that user can retry volume expansion with values that may succeed.
+
+To enable recovery, this proposal relaxes API validation for PVC objects to allow reduction in
+requested size. This KEP also adds a new field called `pvc.Spec.AllocatedResources` to allow resource quota
+to be tracked accurately if user reduces size of their PVCs.
+
+
+## Motivation
+
+Sometimes a user may expand a PVC to a size which may not be supported by underlying storage provider.  Lets say PVC size was 10GB and user expanded it to 500GB but underlying storage provider can not support 500GB volume for some reason. We want to make it easier for user to retry resizing by requesting more reasonable value say - 100GB.
+
+Problem in allowing users to reduce size is - a malicious user can use this feature to abuse the quota system. They can do so by rapidly expanding and then shrinking the PVC. In general - both CSI and in-tree volume plugins have been
+designed to never perform actual shrink on underlying volume, but it can fool the quota system in believing the user is using less storage than he/she actually is. To solve this problem - we propose that although users are allowed to reduce size of their PVC (as long as requested size >= `pvc.Status.Capacity`), quota calculation will only use previously requested higher value. This is covered in more detail in [Implementation](#implementation).
+
+
+
+### Goals
+
+- Allow users to cancel previously issued volume expansion requests, assuming they are not yet successful or have failed.
+- Allow users to retry expansion request with smaller value than original requested size in `pvc.Spec.Resources`, assuming previous requests are not yet successful or have failed.
+
+### Non-Goals
+
+- As part of this KEP we do not intend to actually support shrinking of underlying volume.
+
+## Proposal
+
+As part of this proposal, we are mainly proposing three changes:
+
+1. Relax API validation on PVC update so as reducing `pvc.Spec.Resoures` is allowed, as long requested size is `>=pvc.Status.Capacity`.
+2. Add a new field in PVC api called - `pvc.Spec.AllocatedResources` for purpose of tracking used storage size. This field is only allowed to increase.
+3. Update quota code to use `max(pvc.Spec.Resources, pvc.Spec.AllocatedResources)` when evaluating usage for PVC.
```

Comment:
I dont understand what the difference between `pvc.Spec.Resources` and `AllocatedResources` is. Shouldn't they be the same? In what cases should they differ and what does that mean to me as a user. If `AllocatedResources` is something that changes based on the state of the system should it be in `Status`?

Reply:
`pvc.spec.Resources` is basically allowed to go up or down (as long as it is greater than `pvc.status.capacity`) but `pvc.spec.allocatedResources` is *never* allowed to go down because if it does, it will allow an malicious user to fool the quota system. I think I should document some user stories what about what will happen if we did implement reducing quota size when user retries expansion with a lower value than original request.